packages/salesforcedx-vscode-core/src/services/sdr/componentSetUtils.ts

```diff
@@ -67,3 +67,9 @@ export async function getOrgApiVersion(): Promise<string> {
 
   return apiVersion;
 }
+
+module.exports = {
```

Comment:
q: why are we using module.exports?  I had to google it to understand what it does different (and I learned the other exports in this file are ignored) Using the pattern found in https://github.com/forcedotcom/salesforcedx-vscode/blob/fb6111619ca8c6cbf2dc24100b75c626900dee1a/packages/salesforcedx-utils-vscode/src/workspaces/workspaceUtils.ts#L24-L29 seems like it'd be more consistent with the rest of the code base.

Reply:
The problem I'm trying to solve is when a file exports several functions, with one of the functions being a dependency of the other, and then mocking it.
```
export function a() {
  b()
}

export function b() {
}
```

(note: I don't think the syntax here isn't the issue, as I've also tried the CommnJS style of exporting, as well as the style you referenced in workspaceUtils.ts but still not able to mock b())

In my unit tests for this file, I wish to test a(), and mock b()
(and then later when I test b(), to use the actual impl of b(), but that's a different issue)

I ran into problems with this, and in searching around I found this posting which recommended using `exports.`:
https://medium.com/welldone-software/jest-how-to-mock-a-function-call-inside-a-module-21c05c57a39f

This works w/o issue when I run or debug locally, but fails when using a VSIX file, so I'm looking for a different solution.

(I also tried "solution 2" here: https://dev.to/edwinwong90/how-to-mock-function-that-called-inside-module-7f7 but this too did not work)

Do we have test within our codebase which mock inner file dependencies, like the example above with b() being consumed by a()?

Reply:
As an example (using workspaceUtils.ts as example, first I modified workspaceUtils.ts just to make it easier to flush out a solution:
```
import { workspace, WorkspaceFolder } from 'vscode';

function hasRootWorkspace(ws: typeof workspace = workspace) {
  debugger;
  // this should not get called if it has been mocked

  // return ws && ws.workspaceFolders && ws.workspaceFolders.length > 0;
  return false;
}

function getRootWorkspace(): WorkspaceFolder {
  // return hasRootWorkspace()
  //   ? workspace.workspaceFolders![0]
  //   : ({} as WorkspaceFolder);

  debugger;

  return hasRootWorkspace()
    ? ({ name: 'was true'  } as WorkspaceFolder)
    : ({ name: 'was false' } as WorkspaceFolder);
}

function getRootWorkspacePath(): string {
  return getRootWorkspace().uri ? getRootWorkspace().uri.fsPath : '';
}

export const workspaceUtils = {
  hasRootWorkspace,
  getRootWorkspace,
  getRootWorkspacePath
};
```


Next, I created a corresponding workspaceUtils.test.ts file:
```
// import {
//   workspaceUtils
// } from '../../../src/';

// import {
//   workspaceUtils
// } from '../../../src/workspaces';

import {
  workspaceUtils
} from '../../../src/workspaces/workspaceUtils';


describe('test project paths', () => {
  it('should return a path to the state folder if the project has a root workspace', () => {

    // mock hasRootWorkspace
    const hasRootWorkspaceSpy = jest.spyOn(workspaceUtils, 'hasRootWorkspace');
    hasRootWorkspaceSpy.mockReturnValue(true);

    debugger;

    // then call getRootWorkspace
    const rootWorkspace = workspaceUtils.getRootWorkspace();

    debugger;

    expect(rootWorkspace.name).toBe('was true');

    debugger;
  });
});
```

...but alas, everytime `getRootWorkspace()` is called, the actual impl of `hasRootWorkspace()` is called.

Reply:
gotcha. Good ol' javascript scoping fun :). The issue here is how does func1 call func2 in such a way that we have access to mock it b/c is using the same reference as the test file. 

I think you have a few options.  
1.  Have one function call the other via the grouping object.  So in the above workspaceUtils change `return hasRootWorkspace()` to `return workspaceUtils.hasRootWorkspace()`.
2. Another option (which I always have to exercise and can get hairy) is to use this when calling the other function in the same module ala `return this.hasRootWorkspace()`.  I think this in this case will refer back to the workspaceUtils object giving you the right access for mocking. 
3. You can refactor to have just a single export ```

Reply:
Thanks, option `#1` is working out.  I've updated the PR and the other places where componentSetUtils is used.

@RitamAgrawal and @CristiCanizales I've d/l and installed the VSIX files were were generated from this PR, and after installing I was able to deploy, so things are looking for for this PR.  Please review & test, and let me know if either of you discover any issues.

Reply:
nice breakdown of the problem btw 👍